File: code/trees/tree.py


class TreeNode:
    def __init__(self, data, parent=None, children=[]):
        self.data = data
        self.parent = parent
        self.children = children

    def is_root(self):
        return self.parent is None

    def is_internal(self):
        return len(self.children) > 0

    def is_external(self):
        return not self.is_internal()

    def get_children(self):
        return self.children

    def add_child(self, child):
        child.parent = self
        self.children.append(child)

    def get_num_descendants(self):
        count = 0
        for child in self.children:
            count += 1
            count += child.get_num_descendants()
        return count


class Tree:
    def __init__(self, root=None):
        self.root = root

    def is_empty(self):
        return self.root is None

    def size(self):
        if self.root is not None:
            count = 1 + self.root.get_num_descendants()
            return count
        return 0

File: code/trees/tree.cpp
#include <iostream>
#include <cstdlib>

#define MAX_NUM_CHILDREN 100

using namespace std;

typedef data_t int

typedef struct tree_node
{
  data_t data;
  int num_children = 0;
  struct tree_node* children[MAX_NUM_CHILDREN];
  struct tree_node* parent;
} TreeNode, *TreeNodePtr

TreeNodePtr make_tree_node(data_t data, TreeNodePtr parent)
{
  TreeNodePtr tnp = (TreeNodePtr)malloc(sizeof(TreeNode));
  tnp->data = data;
  tnp->num_chilren = 0;
  tnp->parent = parent;
  memset(tnp->children, 0, sizeof(TreeNodePtr) * MAX_NUM_CHILDREN);
  return tp;
}

void TreeNodePtr add_child(TreeNodePtr parent, TreeNodePtr child)
{
  child->parent = parent;
  if(parent->num_children < MAX_NUM_CHILDREN)
  {
    parent->children[parent->num_children] = child;
    parent->num_children += 1
  }
}

int is_internal(TreeNodePtr tnp)
{
  return (tnp->num_children > 0);
}

int is_external(TreeNodePtr tnp)
{
  return !(is_internal(tnp));
}

int num_descendants(TreeNodePtr tnp)
{
  int count = 0;
  for(int i = 0; i < tnp->num_children; i++)
  {
    count += num_descendants(tnp->children[i]);
  }
  return count;
}

typedef struct
{
  TreeNodePtr root;
} Tree, *TreePtr;

TreePtr create_tree(TreeNodePtr root)
{
  TreePtr tree = (TreePtr)malloc(sizeof(Tree));
  tree->root = root;
  return tree;
}

int is_empty(TreePtr tp)
{
  return tp->root == NULL;
}

int tree_size(TreePtr tp)
{
  if(tp->root == NULL)
    return 0;
  int count = 1 + num_descendants(tp->root);
  return count;
}
